File: src/MagnusAddon/language_services/janome_ex/tokenizing/jn_tokenizer.py
from __future__ import annotations

from typing import final

from autoslot import Slots  # pyright: ignore[reportMissingTypeStubs]
from janome.tokenizer import Token, Tokenizer  # pyright: ignore[reportMissingTypeStubs]
from language_services.janome_ex.tokenizing.jn_parts_of_speech import JNPartsOfSpeech
from language_services.janome_ex.tokenizing.jn_token import JNToken
from language_services.janome_ex.tokenizing.jn_tokenized_text import JNTokenizedText
from sysutils import ex_str, typed


@final
class JNTokenizer(Slots):
    def __init__(self) -> None:
        self._tokenizer = Tokenizer()

    # apparently janome get's confused by some characters, so we replace them with ordinary spaces
    _character_that_confuses_janome_so_we_replace_them_with_ordinary_full_width_spaces = {ex_str.invisible_space, "!", "！"}

    def tokenize(self, text: str) -> JNTokenizedText:
        # apparently janome does not fully understand that invisible spaces are word separators, so we replace them with ordinary spaces since they are not anything that should need to be parsed...
        sanitized_text = text
        for character in self._character_that_confuses_janome_so_we_replace_them_with_ordinary_full_width_spaces:
            sanitized_text = sanitized_text.replace(character, " ")

        # it seems that janome is sometimes confused and changes its parsing if there is no whitespace after the text, so lets add one
        tokens = [typed.checked_cast(Token, token) for token in self._tokenizer.tokenize(sanitized_text)]

        # Create JNToken objects
        jn_tokens = [JNToken(JNPartsOfSpeech.fetch(typed.str_(token.part_of_speech)),  # pyright: ignore[reportAny]
                             typed.str_(token.base_form),  # pyright: ignore[reportAny]
                             typed.str_(token.surface),  # pyright: ignore[reportAny]
                             typed.str_(token.infl_type),  # pyright: ignore[reportAny]
                             typed.str_(token.infl_form),  # pyright: ignore[reportAny]
                             typed.str_(token.reading),  # pyright: ignore[reportAny]
                             typed.str_(token.phonetic),  # pyright: ignore[reportAny]
                             typed.str_(token.node_type),  # pyright: ignore[reportAny]
                             token) for token in tokens]

        # Link tokens with previous/next pointers
        for i in range(len(jn_tokens)):
            if i > 0:
                jn_tokens[i]._previous = jn_tokens[i - 1].weak_ref  # pyright: ignore [reportPrivateUsage]
            if i < len(jn_tokens) - 1:
                jn_tokens[i]._next = jn_tokens[i + 1].weak_ref  # pyright: ignore [reportPrivateUsage]

        return JNTokenizedText(text, tokens, jn_tokens)

File: src/src_dotnet/JAStudio.Core/LanguageServices/JanomeEx/Tokenizing/JNTokenizer.cs
using System.Collections.Generic;
using System.Linq;
using Compze.Utilities.SystemCE;
using MeCab;

namespace JAStudio.Core.LanguageServices.JanomeEx.Tokenizing;

public sealed class JNTokenizer
{
   static readonly LazyCE<JNTokenizer> Instance = new(() => new JNTokenizer());

   public static JNTokenizer GetInstance() => Instance.Value;

   static readonly HashSet<string> CharactersThatMayConfuseTokenizerSoWeReplaceThemWithOrdinaryFullWidthSpaces =
      ["!", "！", "|", "（", "）"];

   /// <summary>The invisible space character (U+200B) used as field separator in the serialized token string.</summary>
   const string FieldSeparator = StringExtensions.InvisibleSpace;

   readonly MeCabTagger _tagger;
   readonly object _lock = new();

   JNTokenizer()
   {
      _tagger = MeCabTagger.Create();
   }

   public JNTokenizeResult Tokenize(string text, string? cachedSerializedTokens = null)
   {
      // The tokenizer does not fully understand that invisible spaces are word separators,
      // so we replace them with a sentinel token since they are not anything that should need to be parsed
      var sanitizedText = text.Replace(StringExtensions.InvisibleSpace, JNToken.SplitterTokenText);

      foreach(var character in CharactersThatMayConfuseTokenizerSoWeReplaceThemWithOrdinaryFullWidthSpaces)
      {
         sanitizedText = sanitizedText.Replace(character, " ");
      }

      // Use cached serialized tokens if available, otherwise tokenize with MeCab
      var serialized = cachedSerializedTokens ?? TokenizeToSerializedString(sanitizedText);

      var jnTokens = ParseSerializedTokens(serialized);

      // Link tokens with previous/next pointers
      for(var i = 0; i < jnTokens.Count; i++)
      {
         if(i > 0)
         {
            jnTokens[i].Previous = jnTokens[i - 1];
         }

         if(i < jnTokens.Count - 1)
         {
            jnTokens[i].Next = jnTokens[i + 1];
         }
      }

      return new JNTokenizeResult(new JNTokenizedText(text, jnTokens), serialized);
   }

   string TokenizeToSerializedString(string text)
   {
      lock(_lock)
      {
         var lines = new List<string>();

         foreach(var node in _tagger.ParseToNodes(text))
         {
            if(node.CharType == 0) continue; // Skip BOS/EOS nodes

            var features = node.Feature.Split(',');

            // IPAdic feature CSV format: POS,sub1,sub2,sub3,inflType,inflForm,baseForm,reading,phonetic
            var partOfSpeech = string.Join(",", features.Take(4));
            var inflType = GetFeatureOrEmpty(features, 4);
            var inflForm = GetFeatureOrEmpty(features, 5);
            var baseForm = GetFeatureOrEmpty(features, 6);
            var reading = GetFeatureOrEmpty(features, 7);
            var phonetic = GetFeatureOrEmpty(features, 8);

            var fields = string.Join(FieldSeparator, [
               partOfSpeech,
               Sanitize(baseForm),
               Sanitize(node.Surface),
               Sanitize(inflType),
               Sanitize(inflForm),
               Sanitize(reading),
               Sanitize(phonetic),
               "" // node_type — not actively used in the codebase
            ]);

            lines.Add(fields);
         }

         return string.Join("\n", lines);
      }
   }

   static string GetFeatureOrEmpty(string[] features, int index) =>
      index < features.Length ? features[index] : "";

   static string Sanitize(string? value) =>
      string.IsNullOrEmpty(value) ? "" : value.Replace("\n", " ").Replace("\r", " ");

   static List<JNToken> ParseSerializedTokens(string serialized)
   {
      if(string.IsNullOrEmpty(serialized))
         return [];

      var lines = serialized.Split('\n');
      var result = new List<JNToken>(lines.Length);

      foreach(var line in lines)
      {
         if(line.Length == 0) continue;

         var fields = line.Split(FieldSeparator);
         // Fields: 0=part_of_speech, 1=base_form, 2=surface, 3=infl_type, 4=infl_form, 5=reading, 6=phonetic, 7=node_type
         var partsOfSpeech = JNPartsOfSpeech.Fetch(fields[0]);

         result.Add(new JNToken(
                       partsOfSpeech,
                       baseForm: fields[1],
                       surface: fields[2],
                       inflectionType: fields[3],
                       inflectedForm: fields[4],
                       reading: fields[5],
                       phonetic: fields[6],
                       nodeType: fields[7]
                    ));
      }

      return result;
   }
}
